pilot/pkg/model/push_context.go

```diff
@@ -604,49 +630,93 @@ func (ps *PushContext) SetDestinationRules(configs []Config) {
 	// Sort by time first. So if two destination rule have top level traffic policies
 	// we take the first one.
 	sortConfigByCreationTime(configs)
+	privateHostsByNamespace := make(map[string][]Hostname, 0)
+	privateCombinedDestinationRuleMap := make(map[string]map[Hostname]*combinedDestinationRule, 0)
 	hosts := make([]Hostname, 0)
-	combinedDestinationRuleMap := make(map[Hostname]*combinedDestinationRule, len(configs))
+	combinedDestinationRuleMap := make(map[Hostname]*combinedDestinationRule, 0)
 
 	for i := range configs {
 		rule := configs[i].Spec.(*networking.DestinationRule)
 		resolvedHost := ResolveShortnameToFQDN(rule.Host, configs[i].ConfigMeta)
-		if mdr, exists := combinedDestinationRuleMap[resolvedHost]; exists {
-			combinedRule := mdr.config.Spec.(*networking.DestinationRule)
-			// we have an another destination rule for same host.
-			// concatenate both of them -- essentially add subsets from one to other.
-			for _, subset := range rule.Subsets {
-				if _, subsetExists := mdr.subsets[subset.Name]; !subsetExists {
-					mdr.subsets[subset.Name] = true
-					combinedRule.Subsets = append(combinedRule.Subsets, subset)
-				} else {
-					ps.Add(DuplicatedSubsets, string(resolvedHost), nil,
-						fmt.Sprintf("Duplicate subset %s found while merging destination rules for %s",
-							subset.Name, string(resolvedHost)))
+		if rule.ConfigScope == networking.ConfigScope_PRIVATE {
+			if _, exist := privateCombinedDestinationRuleMap[configs[i].Namespace]; !exist {
+				privateCombinedDestinationRuleMap[configs[i].Namespace] = map[Hostname]*combinedDestinationRule{}
+			}
+
+			if mdr, exists := privateCombinedDestinationRuleMap[configs[i].Namespace][resolvedHost]; exists {
+				combinedRule := mdr.config.Spec.(*networking.DestinationRule)
+				// we have an another destination rule for same host.
+				// concatenate both of them -- essentially add subsets from one to other.
+				for _, subset := range rule.Subsets {
+					if _, subsetExists := mdr.subsets[subset.Name]; !subsetExists {
+						mdr.subsets[subset.Name] = struct{}{}
+						combinedRule.Subsets = append(combinedRule.Subsets, subset)
+					} else {
+						ps.Add(DuplicatedSubsets, string(resolvedHost), nil,
+							fmt.Sprintf("Duplicate subset %s found while merging destination rules for %s",
+								subset.Name, string(resolvedHost)))
+					}
+
+					// If there is no top level policy and the incoming rule has top level
+					// traffic policy, use the one from the incoming rule.
+					if combinedRule.TrafficPolicy == nil && rule.TrafficPolicy != nil {
+						combinedRule.TrafficPolicy = rule.TrafficPolicy
+					}
```

Comment:
To avoid errors caused by copy paste and repetitive code, I suggest you move the logic into a small function

```combineSingleDestinationRule(combinedDestRuleHosts []Hostname, combinedDestinationRuleMap map[Hostname]*combinedDestinationRule, destRuleConfig config.Config)```

Reply:
SGTM